### PR TITLE
Fix range properties

### PR DESCRIPTION
**Pull request by DorianZheng**
## What have you changed? (mandatory)
Calculate blob index size

## What are the type of the changes? (mandatory)
- Bug fix

## How has this PR been tested? (mandatory)
CI

## Does this PR affect documentation (docs) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Refer to a related PR or issue link (optional)
https://github.com/tikv/tikv/issues/4175

### CODE REVIEW

**Comment by siddontang** [comment]
any test?

**Comment by zhangjinpeng1987** [comment]
@DorianZheng Please fix conflicts.

File: src/util/rocksdb/properties.rs [hunk 1]
```diff
@@ -364,11 +365,17 @@ impl SizePropertiesCollector {
...
     fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
-        if entry_type != DBEntryType::Put {
-            return;
-        }
-        let size = key.len() + value.len();
-        self.index_handle.size += size as u64;
+        let value_size = match entry_type {
+            DBEntryType::Put => value.len() as u64,
+            DBEntryType::BlobIndex => match TitanBlobIndex::decode_from(value) {
+                Ok(index) => index.blob_size,
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/properties.rs]: I think also need to add value.len() even if the type is blob, because value do occupy space.

Comment (zhangjinpeng1987) [reply src/util/rocksdb/properties.rs]: Currently it is about 16 bytes, in the future we may store more informations in it, for example timestamp.

File: src/util/rocksdb/properties.rs [hunk 2]
```diff
@@ -615,15 +622,20 @@ impl RangePropertiesCollector {
 
 impl TablePropertiesCollector for RangePropertiesCollector {
     fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
-        if entry_type != DBEntryType::Put {
-            return;
-        }
+        let value_size = match entry_type {
+            DBEntryType::Put => value.len() as u64,
+            DBEntryType::BlobIndex => match TitanBlobIndex::decode_from(value) {
+                Ok(index) => index.blob_size,
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/properties.rs]: Ditto.

**Comment by huachaohuang** [comment]
@DorianZheng please add some tests.

**Comment by DorianZheng** [comment]
#https://github.com/pingcap/rust-rocksdb/pull/272

File: src/util/rocksdb_util/properties.rs [hunk 3]
```diff
@@ -940,4 +953,64 @@ mod tests {
...
+            // handle "k": size(size = DISTANCE + 2, offset = DISTANCE / 8 * 25 + 11),keys(2,11)
+        ];
+
+        let handles = ["a", "e", "i", "k"];
+
+        let mut rng = rand::thread_rng();
+        let mut collector = RangePropertiesCollector::default();
+        let mut extra_value_size: u64 = 0;
+        for &(k, vlen, count) in &cases {
+            if handles.contains(&k) || rng.gen_range(0, 2) == 0 {
```
Comment (huachaohuang) [inline src/util/rocksdb_util/properties.rs]: Is it necessary to add randomness here?

File: src/util/rocksdb_util/properties.rs [hunk 4]
```diff
@@ -940,4 +953,64 @@ mod tests {
...
+                for _ in 0..count {
+                    let v = vec![0; vlen as usize - extra_value_size as usize];
+                    extra_value_size = 0;
+                    collector.add(k.as_bytes(), &v, DBEntryType::Put, 0, 0);
+                }
+            } else {
+                let mut blob_index = TitanBlobIndex::default();
+                for _ in 0..count {
```
Comment (huachaohuang) [inline src/util/rocksdb_util/properties.rs]: Seems all counts are 1, can we just remove them?

File: src/util/rocksdb_util/properties.rs [hunk 5]
```diff
@@ -364,11 +365,17 @@ impl SizePropertiesCollector {
 
 impl TablePropertiesCollector for SizePropertiesCollector {
     fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
-        if entry_type != DBEntryType::Put {
-            return;
-        }
-        let size = key.len() + value.len();
-        self.index_handle.size += size as u64;
+        let value_size = match entry_type {
```
Comment (huachaohuang) [inline src/util/rocksdb_util/properties.rs]: Maybe put this piece of code into a `get_entry_size` function?

File: src/util/rocksdb_util/properties.rs (hunk 3 above)
Comment (DorianZheng) [reply src/util/rocksdb_util/properties.rs]: doesn't harm

Comment (huachaohuang) [reply src/util/rocksdb_util/properties.rs]: It harms because the reviewer need to think about it and find out that it is unnecessary.

Comment (DorianZheng) [reply src/util/rocksdb_util/properties.rs]: no, it's necessary

File: src/util/rocksdb_util/properties.rs [hunk 6]
```diff
@@ -11,6 +11,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std;
```
Comment (overvenus) [inline src/util/rocksdb_util/properties.rs]: You need to `use std;` in 2018 edtion.

File: src/util/rocksdb_util/properties.rs [hunk 7]
```diff
@@ -364,11 +377,8 @@ impl SizePropertiesCollector {
 
 impl TablePropertiesCollector for SizePropertiesCollector {
     fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
-        if entry_type != DBEntryType::Put {
-            return;
-        }
-        let size = key.len() + value.len();
-        self.index_handle.size += size as u64;
+        let size = key.len() as u64 + get_entry_size(value, entry_type).unwrap();
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb_util/properties.rs]: When `entry_type` is `DBEntryType::Delete`, `get_entry_size` will return Err(()) which will cause panic here.

File: src/util/rocksdb_util/properties.rs [hunk 8]
```diff
@@ -615,15 +625,11 @@ impl RangePropertiesCollector {
...
-        if entry_type != DBEntryType::Put {
-            return;
-        }
-
         // keys
         self.cur_offsets.keys += 1;
         // size
-        let size = key.len() + value.len();
-        self.cur_offsets.size += size as u64;
+        let size = key.len() as u64 + get_entry_size(value, entry_type).unwrap();
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb_util/properties.rs]: Ditto.

File: src/util/rocksdb_util/properties.rs [hunk 9]
```diff
@@ -940,4 +947,60 @@ mod tests {
...
+        for &(k, vlen) in &cases {
+            if handles.contains(&k) || rng.gen_range(0, 2) == 0 {
+                let v = vec![0; vlen as usize - extra_value_size as usize];
+                extra_value_size = 0;
+                collector.add(k.as_bytes(), &v, DBEntryType::Put, 0, 0);
+            } else {
+                let mut blob_index = TitanBlobIndex::default();
+                blob_index.blob_size = vlen - extra_value_size;
+                let v = blob_index.encode();
+                extra_value_size = v.len() as u64;
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb_util/properties.rs]: I'm confused by the way of calculating `extra_value_size`, could you give some comments?

File: src/util/rocksdb_util/properties.rs [hunk 10]
```diff
@@ -42,6 +44,17 @@ const PROP_RANGE_INDEX: &str = "tikv.range_index";
...
 const PROP_KEYS_INDEX_DISTANCE: u64 = 40 * 1024;
 
+fn get_entry_size(value: &[u8], entry_type: DBEntryType) -> std::result::Result<u64, ()> {
+    match entry_type {
+        DBEntryType::Put => Ok(value.len() as u64),
+        DBEntryType::BlobIndex => match TitanBlobIndex::decode(value) {
+            Ok(index) => Ok(index.blob_size + value.len() as u64),
+            Err(_) => Err(()),
+        },
+        _ => Err(()),
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb_util/properties.rs]: How about s/Err(())/Ok(0)?

File: src/util/rocksdb_util/properties.rs [hunk 11]
```diff
@@ -42,6 +44,17 @@ const PROP_RANGE_INDEX: &str = "tikv.range_index";
 const PROP_SIZE_INDEX_DISTANCE: u64 = 4 * 1024 * 1024;
 const PROP_KEYS_INDEX_DISTANCE: u64 = 40 * 1024;
 
+fn get_entry_size(value: &[u8], entry_type: DBEntryType) -> std::result::Result<u64, ()> {
+    match entry_type {
+        DBEntryType::Put => Ok(value.len() as u64),
+        DBEntryType::BlobIndex => match TitanBlobIndex::decode(value) {
+            Ok(index) => Ok(index.blob_size + value.len() as u64),
+            Err(_) => Err(()),
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb_util/properties.rs]: Log some info when decode block index failed.

File: src/util/rocksdb_util/properties.rs (hunk 7 above)
Comment (huachaohuang) [reply src/util/rocksdb_util/properties.rs]: In that case, maybe it's more straight forward to put the code back here :(

File: src/util/rocksdb_util/properties.rs (hunk 3 above)
Comment (huachaohuang) [reply src/util/rocksdb_util/properties.rs]: Can you explain the test? I'm a bit confusing.

**Comment by zhangjinpeng1987** [comment]
Ping @DorianZheng

**Comment by DorianZheng** [comment]
@zhangjinpeng1987 PTAL

File: src/util/rocksdb_util/properties.rs (hunk 6 above)
Comment (breeswish) [inline src/util/rocksdb_util/properties.rs]: We don't need this line any more.

File: src/util/rocksdb_util/properties.rs (hunk 10 above)
Comment (breeswish) [reply src/util/rocksdb_util/properties.rs]: How about `Option<u64>`?

**Pull request merged**